### PR TITLE
Data manager to fetch RefSeq FASTA

### PR DESCRIPTION
**Pull request by pvanheus**
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

FOR REVIEWER:
* [ ] .shed.yml file ok
    - [ ] Toolshed user `iuc` has access to associated toolshed repo(s)
* [ ] Indentation is correct (4 spaces)
* [ ] Tool version/build ok
* [ ] `<command/>`
  - [ ] Text parameters, input and output files `'single quoted'`
  - [ ] Use of `<![CDATA[ ... ]]>` tags
  - [ ] Parameters of type `text` or having `optional="true"` attribute are checked with `if str($param)` before being used
* [ ] Data parameters have a `format` attribute containing datatypes recognised by Galaxy
* [ ] Tests
  - [ ] Parameters are reasonably covered
  - [ ] Test files are appropriate
* [ ] Help
  - [ ] Valid restructuredText and uses `<![CDATA[ ... ]]>` tags
* [ ] Complies with other best practice in [Best Practices Doc](http://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html)

### CODE REVIEW

**Comment by pvanheus** [comment]
Add a new data manager that allows fetching FASTA data from RefSeq and adding to the `all_fasta` table. Further PRs will be added to ensure that it complies with checklist items.

File: data_managers/data_manager_fetch_refseq/data_manager/fetch_refseq.xml [hunk 1]
```diff
@@ -0,0 +1,87 @@
+<tool id="data_manager_fetch_refseq" name="RefSeq data manager" version="0.0.18" tool_type="manage_data">
+    <description>Fetch FASTA data from NCBI RefSeq and update all_fasta data table</description>
+    <requirements>
+        <requirement type="package" version="3">python</requirement>
+    </requirements>
+    <command detect_errors="aggressive"><![CDATA[
```
Comment (mvdbeek) [inline data_managers/data_manager_fetch_refseq/data_manager/fetch_refseq.xml]: `exit_code` seems more appropriate here, it's a python script so I assume it's well-behaved.

File: data_managers/data_manager_fetch_refseq/.shed.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+categories:
+- Data Managers
+description: Fetch FASTA data from NCBI RefSeq and update all_fasta data table
+long_description: |
+  This data manager fetches FASTA format collections of proteins, nucleotides (genomic DNA) and RNA
+  from NCBI's RefSeq data collection.
+name: data_manager_fetch_refseq
+owner: iuc
+remote_repository_url: https://github.com/pvanheus/refseq_fasta_data_manager
```
Comment (mvdbeek) [inline data_managers/data_manager_fetch_refseq/.shed.yml]: This should be `https://github.com/galaxyproject/tools-iuc/tree/master/data_managers/data_manager_fetch_refseq`

File: data_managers/data_manager_fetch_refseq/data_manager/fetch_refseq.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+import os
+import os.path
+import sys
+from datetime import date
+from multiprocessing import Process, Queue
+
+import requests
+
+try:
+    from io import StringIO
```
Comment (mvdbeek) [inline data_managers/data_manager_fetch_refseq/data_manager/fetch_refseq.py]: If you want to make this py2-py3 compatible you should drop the python 3 requirement (and change python3 to python).

File: data_managers/data_manager_fetch_refseq/data_manager/fetch_refseq.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            for i, mol_type in enumerate(mol_types):
+                assert mol_type in fasta_files[i], "Filename does not contain expected mol_type ({}, {})".format(mol_type, fasta_files[i])
+                unique_key = 'refseq_' + division_name + '.' + release_num + '.' + mol_type  # note: this is now same as dbkey
+                dbkey = unique_key
+                desc = 'RefSeq ' + division_name + ' Release ' + release_num + ' ' + mol_type + ' (' + today_str + ')'
+                path = os.path.join(output_directory, fasta_files[i])
+                _add_data_table_entry(data_manager_dict=data_manager_dict,
+                                      data_table_entry=dict(value=unique_key, dbkey=dbkey, name=desc, path=path),
+                                      data_table_name='all_fasta')
+            open(args.galaxy_datamanager_filename, 'wb').write(json.dumps(data_manager_dict).encode())
```
Comment (mvdbeek) [inline data_managers/data_manager_fetch_refseq/data_manager/fetch_refseq.py]: You can just open the file in `w` mode and drop the additional encode. Also can you add `sort_keys=True` to json.dumps ?

File: data_managers/data_manager_fetch_refseq/data_manager/fetch_refseq.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        'rna': 'rna.fna.gz'
+    }
+    assert division_name in valid_divisions, "Unknown division name ({})".format(division_name)
+    for mol_type in mol_types:
+        assert mol_type in ending_mappings, "Unknown molecule type ({})".format(mol_type)
+    if not os.path.exists(output_directory):
+        os.mkdir(output_directory)
+    release_num_file = base_url + 'RELEASE_NUMBER'
+    r = requests.get(release_num_file)
+    release_num = str(int(r.text.strip()))
```
Comment (mvdbeek) [inline data_managers/data_manager_fetch_refseq/data_manager/fetch_refseq.py]: Could this just be `.text.strip()` ?

File: data_managers/data_manager_fetch_refseq/data_manager/fetch_refseq.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+import functools
+import gzip
+import json
+import os
+import os.path
+import sys
+from datetime import date
+from multiprocessing import Process, Queue
+
+import requests
```
Comment (mvdbeek) [inline data_managers/data_manager_fetch_refseq/data_manager/fetch_refseq.py]: requests is not part of the stdlib, you need to add it to the requirements

**Comment by pvanheus** [comment]
Thanks @mvdbeek the idea of creating a "data_source" tool is interesting. Effectively the problem most of the Python code here addresses is downloading and concatenating the individual files from RefSeq into a single file and then associating the RefSeq release information (division and release number) with that file. If this was a data_source tool then the output would go to the history which would then form input for https://github.com/galaxyproject/tools-iuc/tree/master/data_managers/data_manager_fetch_genome_dbkeys_all_fasta. I'm not sure how the information on release number etc would be preserved in this process. Would it not have to be entered again as the history item (a FASTA file) was used as the source for the data manager?

Or do you mean making an additional tool for non-admin users to use? That should be straightforward - I could abstract some of the code out into a module and share code between the two tools.

As to the .len files, I think those are used for genome contig lengths, whereas this deals with the part of RefSeq that relates to gene / protein / transcript sequences. Are .len files still relevant? If so I can create them.

**Comment by bernt-matthias** [comment]
would be nice to have a md5 test

**Comment by pvanheus** [comment]
> would be nice to have a md5 test

It would be nice @bernt-matthias but I don't see a source of MD5 checksums on the NCBI FTP site.

**Comment by pvanheus** [comment]
> On NCBI there are some: ftp://ftp.ncbi.nlm.nih.gov/blast/db/

Yes those are for the BLAST databases. I agree that NCBI should have checksums for RefSeq. Unfortunately changing NCBI practice is beyond the scope of this PR.

@mvdbeek do you have any further feedback on my comments? Specifically on the question of this as a tool and of the `.len` files.

**Comment by mvdbeek** [comment]
I don't know, i guess it's OK to not have .len files, there are not that many tools that use them anyway. On the other hand you might also want to know the number of amino acids in a protein, right ? I don't have a strong opinion on this.

**Comment by pvanheus** [comment]
@mvdbeek I've had a look at other data managers and they seem to associated .len files with the __dbkey__ table. So I think they are not appropriate for this tool. So I think this is ready for a merge.

**Comment by mvdbeek** [comment]
Thanks @pvanheus !

**Pull request merged**